iroh-willow/src/net.rs

```diff
@@ -0,0 +1,671 @@
+        Role::Betty => conn.accept_bi().await?,
+    };
+    control_send_stream.set_priority(i32::MAX)?;
+    debug!("control channel ready");
+
+    let initial_transmission =
+        exchange_commitments(&mut control_send_stream, &mut control_recv_stream).await?;
+    debug!("exchanged commitments");
+
+    let (control_send, control_recv) = spawn_channel(
+        &mut tasks,
+        Channel::Control,
+        CHANNEL_CAP,
+        CHANNEL_CAP,
+        Guarantees::Unlimited,
+        control_send_stream,
+        control_recv_stream,
+    );
+
+    let (logical_send, logical_recv) = open_logical_channels(&mut tasks, conn, our_role).await?;
+    debug!("logical channels ready");
+    let channels = Channels {
+        send: ChannelSenders {
+            control_send,
+            logical_send,
+        },
+        recv: ChannelReceivers {
+            control_recv,
+            logical_recv,
+        },
+    };
+    let handle = actor
+        .init_session(peer, our_role, initial_transmission, channels, init)
+        .await?;
+
+    Ok(SessionHandle { handle, tasks })
+}
+
+#[derive(Debug)]
+pub struct SessionHandle {
+    handle: actor::SessionHandle,
+    tasks: JoinSet<anyhow::Result<()>>,
+}
+
+impl SessionHandle {
+    /// Close the session gracefully.
+    ///
+    /// After calling this, no further protocol messages will be sent from this node.
+    /// Previously queued messages will still be sent out. The session will only be closed
+    /// once the other peer closes their senders as well.
+    pub fn close(&self) {
+        self.handle.close()
+    }
+
+    /// Wait for the session to finish.
+    ///
+    /// Returns an error if the session failed to complete.
+    pub async fn join(&mut self) -> anyhow::Result<()> {
+        let session_res = self.handle.on_finish().await;
+        let net_tasks_res = join_all(&mut self.tasks).await;
+        match session_res {
+            Err(err) => Err(err.into()),
+            Ok(()) => net_tasks_res,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("missing channel: {0:?}")]
+struct MissingChannel(LogicalChannel);
+
+async fn open_logical_channels(
+    join_set: &mut JoinSet<anyhow::Result<()>>,
+    conn: Connection,
+    our_role: Role,
+) -> anyhow::Result<(LogicalChannelSenders, LogicalChannelReceivers)> {
+    let cap = CHANNEL_CAP;
+    let channels = LogicalChannel::all();
+    let mut channels = match our_role {
+        // Alfie opens a quic stream for each logical channel, and sends a single byte with the
+        // channel id.
+        Role::Alfie => {
+            channels
+                .map(|ch| {
+                    let conn = conn.clone();
+                    async move {
+                        let (mut send, recv) = conn.open_bi().await?;
+                        send.write_u8(ch.id()).await?;
+                        trace!(?ch, "opened bi stream");
+                        Result::<_, anyhow::Error>::Ok((ch, Some((send, recv))))
```

Comment:
you can write `anyhow::Ok` as a shorthand

Reply:
Worked in most places but not all, not sure why.